### ISSUE TITLE
Slow SQL query, WebUI time-out

### ISSUE THREAD

**Issue by jiri-novacek** [Sonerezh/sonerezh#156]
After clicking on "Albums" tab i get timeout error:
504 Gateway Time-out
The server didn't respond in time. 

After investigating a little bit in MySQL I determined that there are two very slow queries executed (each takes ~90 s)

```
# Query_time: 93.026232  Lock_time: 0.000164 Rows_sent: 6  Rows_examined: 190269261
SELECT `Song`.`id`, `Song`.`band`, `Song`.`album`, `Song`.`cover` FROM `sonerezh`.`songs` AS `Song`   WHERE  (`Song`.`id`, `Song`.`album`) IN (SELECT MIN(`subsong`.`id`), `subsong`.`album` FROM `sonerezh`.`songs` AS subsong   WHERE 1 = 1  GROUP BY `subsong`.`album`)    ORDER BY `Song`.`created` DESC  LIMIT 6;
```

```
# Query_time: 91.643405  Lock_time: 0.000095 Rows_sent: 1  Rows_examined: 190269255
SELECT COUNT(*) AS `count` FROM `sonerezh`.`songs` AS `Song`   WHERE  (`Song`.`id`, `Song`.`album`) IN (SELECT MIN(`subsong`.`id`), `subsong`.`album` FROM `sonerezh`.`songs` AS subsong   WHERE 1 = 1  GROUP BY `subsong`.`album`);
```

Understandably WebUI is not waiting long enough for this query and times-out.
Also there is 100% cpu usage while DB is processing these queries.

My library contains:
796 Artists
997 Albums
13959 Songs 

Is this known issue? Is there something wrong with my MySQL? 
I have read that "Sonerezh is not designed yet to manage millions and millions records." but this is just a few thousands...

Other than this my instance is working perfectly and when I had less items in library there was no problem with "Albums" view either.

Thanks for great project, I am looking forward to future versions

**Comment by jiri-novacek** [comment]
After further "troubleshooting"...
Queries and DB structure of Sonerezsh is somewhat crazy, thats why it takes 90s (on Xeon with plenty of RAM) to get something as simple as last 6 added albums from DB.

IMHO it is design flaw, but there is Query Cache in MySQL, which should work as "work-around" for this suboptimal design - at least slow query would be processed just once and then result is cached so next queries takes no time at all (until query cache is flushed/overwritten). Problem is, in my case, that this caching mechanism is not working.

So I investigated why it is not caching (my query cache is enabled and properly configured). After some time I found out that caching works (i can switch between views immediatelly) only until I play any song which has not been played before. After I play any song for the first time, cached query is abandoned and MySQL spends 2x90 s on album sorting and filtering queries. Cache query is abandoned because for every record in "songs" table in DB, there is column "source_path" and column "path". For every newly indexed file there is normal file path in "source_path" and there is just NULL in "path". After song has been played once (first time) value in "path" becomes same filepath as in "source_path". This change in DB invalidates cached queries because data has changed and system needs to reevaluate whole query - spending another three minutes sorting and filtering.

So, as another workaround, I copied everything from "source_path" to "path", so no mater how many new songs I play for the first time, database stays the same (until I add new files and start reindexing) so the cache is valid and I can listen to anything and WebUI is fully working.

SQL query I used to populate "path" with values from "source_path":
`UPDATE songs SET path = source_path`

**Comment by lGuillaume124** [comment]
Hi,
We test Sonerezh with a collection with at least 1228 artists, 693 albums and 11800 tracks. We never had this kind of problem, and it works on Intel Atom CPU D2550 1.86GHz.

The ``path`` field is filled when you play a song and it may change if the track need a re-encoding or not. For example if I have a file named "My-song.flac", the DB record could look like:

```sql
MariaDB [sonerezh]> SELECT `path`, `source_path` FROM `songs` WHERE `source_path` LIKE '%My-song.flac%';
+------+--------------------------------------------------+
| path | source_path                                      |
+------+--------------------------------------------------+
| NULL | /home/guillaume/Music/My-song.flac               |
+------+--------------------------------------------------+
```
If I play this song, Sonerezh tries to convert it to MP3 (according to my settings), and the ``path`` field become:

```sql
+-----------------------------------------------------+--------------------------------------------------+
| path                                                | source_path                                      |
+-----------------------------------------------------+--------------------------------------------------+
| /var/www/webroot/private/app/tmp/20160208214249.mp3 | /home/guillaume/Music/My-song.flac               |
+-----------------------------------------------------+--------------------------------------------------+
```
It could be improved, so don't hesitate to share your suggestions with us here :)

**Comment by jiri-novacek** [comment]
OK, your Atom performance vs my Xeon didn't make any sense to me, so I have upgraded from MySQL 5.5.47 to latest MariaDB and problem is gone. Those same queries are now instant (~0.01s), without caching (overrided for testing with SELECT SQL_NO_CACHE.....) with same tables and no changes in the system, only drop-in replacement (upgrade) of DB engine.

I still think that sonerezh should start using DB in more "DB way" and less "one big excel table" way, but I am not DB expert, nor PHP programmer, so you don't have to take me seriously. 

Maybe one low-hanging improvement - don't write transcoded path when there is no transcoding at all. It just violates caching of table. 
Maybe don't calculate album counts and last couple of albums every time there is jump between tabs, count it during reindexing of music library and remember it (it can't change until next reindex). 
But seeing that it would save 0.02s with MariaDB I am not sure these changes are necessary.

Thanks to this project I have learned a thing or two about MySQL, slow queries debugging, and power of automatic optimisations of queries by DB engines :)

Thanks for your time

**Comment by lGuillaume124** [comment]
We started to design the Sonerezh's database with several tables (like songs, albums, bands, artists...), in a "more DB way", just like you said. But the SQL queries and relations between the tables were quickly too complex and difficult to be maintained... So we abandoned this structure in favor of a more simple one. Only one table named "songs".


However, you are right about the column "path", there is certainly something to change.
